docs(cn): translate content/docs/hooks-state.md into Chinese (#16)
reactjs/zh-hans.reactjs.org: content/docs/hooks-state.md

---
id: hooks-state
title: 使用 State Hook
permalink: docs/hooks-state.html
next: hooks-effect.html
prev: hooks-overview.html
---

> *Hook* 是一个即将到来的特性，它让你不用写 class 也可以使用 state 和其他 React 特性。他们现在在React v16.8.0-alpha.1 中。

Reviewer: 中英文之间漏了个空格；建议意译，准确的前提下让语句更加通顺。

```suggestion
*Hook* 是一个即将到来的特性，它让你不用写 class 也可以使用 state 和其他 React 特性。现在可以在 React v16.8.0-alpha.1 版本中使用它。
```

prev: hooks-overview.html
---

*Hook* 是一个即将到来的特性，它让你不用写 class 也可以使用 state 和其他 React 特性。他们现在在React v16.8.0-alpha.1 中。

> [上一页](/docs/hooks-intro.html)用这个例子介绍了 Hook:

Reviewer 2: ```suggestion
[上一页](/docs/hooks-intro.html)用这个例子介绍了 Hook：
```

```js{4-5}
import { useState } from 'react';

function Example() {

>   // 声明一个新的 state 变量：count

Reviewer 3: 声明一个叫 "count" 的变量

}
```

> 起始 state 为 `{ count: 0 }`，当用户点击按钮后，我们通过调用 `this.setState()` 来增加 `state.count`。我们将在整个页面中使用这个 class 的代码片段。

Reviewer: 这段的大意是说本章后文所有内容都将围绕这个范例进行讲解，之前没体现出来这部分意思。我这个感觉也不太通顺，可以斟酌一下。

```suggestion
state 初始值为 `{ count: 0 }` ，当用户点击按钮后，我们通过调用 `this.setState()` 来增加 `state.count`。后文中我们将使用这个 class 的代码片段做范例。
```

>注意
>

> >你可能想知道为什么我们在这里使用一个计数器例子而不一个更实际的示例。这可以帮助我们将注意力集中到 API 身上，因为我们还在迈出使用 Hook 的第一步。

Reviewer: ```suggestion
>你可能想知道为什么我们在这里使用一个计数器例子而不一个更实际的示例。因为我们还只是初步接触 Hook ，这可以帮助我们将注意力集中到 API 本身。
```

## Hook 和函数定义组件

复习一下， React 的函数定义组件是这样的：

```js
const Example = (props) => {

>   // 你可以在这使用 Hook

Reviewer 2: ```suggestion
  // 你可以在这使用 Hooks
```

  return <div />;
}
```

或是这样：

```js
function Example(props) {

>   // 你可以在这使用 Hook

Reviewer 2: ```suggestion
  // 你可以在这使用 Hooks
```

  return <div />;
}
```

> 你之前可能已经知道这些是“无状态组件”。我们现在介绍的是在这些基础上使用 React State 的能力，所以我们更喜欢叫他"函数定义组件"。

Reviewer 3: state 小写。

> 你之前可能已经知道这些是“无状态组件”。我们现在介绍的是在这些基础上使用 React state 的能力，所以我们更喜欢叫他"函数定义组件"。

Reviewer: ```suggestion
你之前可能把它们叫做“无状态组件”。但现在我们为它们引入了使用 React state 的能力，所以我们更喜欢叫它"函数定义组件"。
```

> Hook 在 class 内部是**不**起作用的。但你可以使用它们来代替编写 class。

Reviewer: ```suggestion
Hook 在 class 内部是**不**起作用的。但你可以使用它们来取代 class 。
```

> ## Hook是什么？

Reviewer 3: Hook 后面空格

> 我们的新示例首先从 React 导入 `useState` Hook

Reviewer: ```suggestion
在下面的新示例中，我们首先从 React 导入 `useState` Hook ：
```

}
```

状态初始化 `{ count: 0 }`,当用户点击按钮后，`this.setState()`被调用来增加 `state.count`。我们在整个页面都会使用到这一段代码。

>注意
>
>你可能会想知道为什么我们要用一个计数器例子而不使用一个更加实际的应用。这是为了帮助我们关注 API 本身，因为我们还在迈出使用 Hooks 的第一步。

> ## Hooks 和函数式组件

Reviewer 4: ```suggestion
## Hooks 和函数定义组件
```
FYI: https://github.com/reactjs/zh-hans.reactjs.org/issues/2
其实我觉得那个“函数定义组件”翻译不是很好，可以讨论一下，然后大家都统一就好了。

复习一下， React 的函数组件是这个样子的：

```js
const Example = (props) => {
  // 你可以在这用Hooks
  return <div />;
}
```

或是这样：

```js
function Example(props) {
  // 你可以在这用Hooks
  return <div />;
}
```

> 你之前可能听过另一个概念"无状态组件"，我们现在介绍的是在"无状态组件"基础上使用 React 的 State ,所以我们更愿意叫他"函数组件"。

Reviewer 4: ditto。关于 function component 术语的翻译需要统一

> Hooks 在 class 内部是 **无效的**。但你可以在不写 class 情况下使用它。

Reviewer 4: ```suggestion
Hooks 在 class 内部是 **无效的**。但你可以在不写 class 的情况下使用它。
```

}
```

> **Hook 是什么？** Hook 是一个特殊的函数，它可以让你 `钩住` React 的特性。例如，`useState` 是一个让你添加 React State 到函数定义组件的 Hook。稍后我们将学习其他 Hook。

Reviewer 3: state 小写

Reviewer 3: "hook into" 感觉翻译成 “钩入” 好一点。

《Hooks at a Glance》里有一句

> Hooks are functions that let you “hook into” React state and lifecycle features from function components. 

“钩入”有加入的意思，“勾住”就有点怪。

> **Hook 是什么？** Hook 是一个特殊的函数，它可以让你 `钩入` React 的特性。例如，`useState` 是一个让你添加 React state 到函数定义组件的 Hook。稍后我们将学习其他 Hook。

Reviewer: ```suggestion
**Hook 是什么？** Hook 是一个特殊的函数，它可以让你 `钩入` React 的特性。例如，`useState` 是一个让你添加 React state 到函数定义组件的 Hook。稍后我们还会接触更多 Hook。
```

> **什么时候我会用 Hook？** 如果你在写一个函数组件并意识到需要向其添加一些状态，如果是以前的话你必须把它转化为一个 class。现在你可以在现有的函数组件中使用 Hook。我们现在就去做！

Reviewer 4: ```suggestion
**什么时候我会用 Hook？** 如果你在写一个函数组件并意识到需要向其添加一些 state，如果是以前的话你必须把它转化为一个 class。现在你可以在现有的函数组件中使用 Hook。我们现在就去做！
```

>注意：
>

> >关于在组件中，有什么地方能使用 Hook，有什么地方不能使用 Hook，有一些特殊的规则。我们将在 [Hook 规范](/docs/hooks-rules.html)中学习它们。

Reviewer: ```suggestion
>在组件中有些特殊的规则，规定什么地方能使用 Hook，什么地方不能使用。我们将在 [Hook 规范](/docs/hooks-rules.html)中学习它们。
```

}
```

> **Hook是什么？** Hook 是一个特殊的函数，它可以让你接入到 React 的特性。例如， `useState` 是一个让你添加 React State 到函数组件的钩子。之后我们将学习其他的钩子函数。

Reviewer 4: Hook 不要翻译成勾子函数，因为上文都直接叫 Hook 了。还有一点，勾子函数容易与生命周期函数的概念混淆，因为那一些也叫勾子函数。

Reviewer 4: Hook 可以不用翻译，直接叫 Hook

> **什么时候我会用到Hook？** 如果你正在写一个函数组件并且你想要把状态加入到组件里面，如果是以前的话你必须把函数转化成class的形式。现在你可以在现有的函数组件内部使用钩子函数。让我们立即行动！

Reviewer 4: Hook 前加一个空格

Reviewer 4: class 两边空格

Reviewer 4: 同上，就不要翻译成“勾子函数”了

>注意：
>

> >这里有一些在组件内部能使用 Hooks 和不能使用 Hooks 的规则。详情请访问[Hooks使用规则](/docs/hooks-rules.html)。

Reviewer 4: ```suggestion
>这里有一些在组件内部能使用 Hooks 和不能使用 Hooks 的规则。详情请访问[Hooks 使用规则](/docs/hooks-rules.html)。
```

Reviewer 4: 空格问题

  }
```

> 在函数组件中，我们没有 `this`，所以我们不能分配或读取 `this.state`。我们直接在组件中调用 `useState` Hook：

Reviewer 4: ```suggestion
在函数定义组件中，我们没有 `this`，所以我们不能分配或读取 `this.state`。我们直接在组件中调用 `useState` Hook：
```
那个术语的规范文档叫做“函数定义组件”

```js{4,5}
import { useState } from 'react';

function Example() {

>   // 声明一个新 state 变量：count

Reviewer 3: ```suggestion
  // 声明一个叫 "count" 的变量
```

  const [count, setCount] = useState(0);
```

> **调用 `useState` 方法做了什么?** 它定义一个 “state 变量”。我们的变量叫 `count`， 但我们可以把它叫做任意的东西，像 `banana`。这是一种在函数调用之间保存一些值的方式—— `useState` 是一种新方法，它和 class 里面的 `this.state` 提供的功能完全相同。一般来说，在函数退出后变量就就会"消失"，但 state 变量是被 React 保留的。

Reviewer 3: 调用 `useState` 方法**时**做了什么?

或者

`useState` 方法做了什么?

Reviewer 3: > 但 state 变量是被 React 保留的

但 state 变量会被 React 保留

> **我们应该传递哪些参数给 `useState`？** `useState()` 方法里面唯一的参数就是初始 state。与 class 不一样，state 可以不是对象。我们可以保留一个数字或字符串，如果这是我们所需要的。在我们示例中，我们只要一个数字来记录用户点击次数，所以我们传了 `0` 作为变量的初始 state。（如果我们想要在 state 中存储两个不同的变量，只需调用 `useState()`两次即可。）

Reviewer 3: > 与 class 不一样，state 可以不是对象。我们可以保留一个数字或字符串，如果这是我们所需要的。

与 class 不同，我们可以按照需要使用数字或字符串，而不一定是要一个对象。

Reviewer 3: > 在我们示例中，我们只要一个数字来记录用户点击次数

在我们的示例中，只要一个数字来记录用户点击次数

Reviewer 3: > 所以我们传了 `0` 作为变量的初始 state

这个 state 不是特指，感觉翻译下比较好。

所以我们传了 `0` 作为变量的初始状态。

> **`useState`方法的返回值是什么？** 它返回一对值：当前 state 和一个更新 state 的函数。这就是我们写 `const [count, setCount] = useState()` 的原因。这和 class 里面 `this.state.count` 和 `this.setState` 类似，唯一区别就是你把它们配成了对。如果你不熟悉我们使用的语法，我们可以返回到[这页的底部](/docs/hooks-state.html#tip-what-do-square-brackets-mean)。

Reviewer 3: useState 后面空格

Reviewer 3: `这和 class 里面 `this.state.count` 和 `this.setState` 类似`

其中一个"和"可以改成"跟"

Reviewer 3: > 唯一区别就是你把它们配成了对

唯一区别是你可以成对地获取它们。

Reviewer 3: > 我们可以返回到[这页的底部]

我们会在这页的底部介绍它。

既然我们知道了 `useState` 的作用，我们的示例应该更容易理解了：

```js{4,5}
import { useState } from 'react';

function Example() {

>   // 声明 state 变量：count

Reviewer 3: ```suggestion
  // 声明一个叫 "count" 的 state 变量
```

  const [count, setCount] = useState(0);
```

> 我们声明了一个叫 `count`的 state 变量，然后把它设为 `0`。React 会在重新渲染的时记住它当前的值，并且提供最新的值给我们的函数。如果我想要更新当前的 `count`，我们可以调用 `setCount`。

Reviewer 3: count 后面空格

Reviewer 3: ```diff
- React 会在重新渲染的时记住它当前的值
+ React 会在重复渲染时记住它当前的值
```

Reviewer 3: ```diff
- 如果我想要更新当前的 `count`，我们可以调用 `setCount`。
+ 我们可以通过调用 `setCount` 来更新当前的 `count`。
```

>注意
>
>你可能想知道：为什么叫 `useState` 而不叫 `createState`?
>

> >"Create"可能不是很准确，因为 state 只在组件首次渲染的时候被创建。在下一次重新渲染时，`useState` 给我们当前的 state。否则 它就不是 “state”了！这也是 Hook的名字*总是*以`use`开头的一个原因。我们将在后面的 [Hook 规则](/docs/hooks-rules.html)中了解原因。

Reviewer 4: ```suggestion
>"Create"可能不是很准确，因为 state 只在组件首次渲染的时候被创建。在下一次重新渲染时，`useState` 返回给我们当前的 state。否则 它就不是 “state”了！这也是 Hook 的名字*总是*以 `use` 开头的一个原因。我们将在后面的 [Hook 规则](/docs/hooks-rules.html)中了解原因。
```

> >"Create"可能不是很准确，因为 state 只在组件首次渲染的时候被创建。在下一次重新渲染时，`useState` 返回给我们当前的 state。否则 它就不是 “state”了！这也是 Hook 的名字*总是*以 `use` 开头的一个原因。我们将在后面的 [Hook 规范](/docs/hooks-rules.html)中了解原因。

Reviewer 3: Create 后面空格

Reviewer 3: “否则”后面多了个空格

## 读取 State

当我们想在 class 中显示当前的 count，我们读取 `this.state.count`：

```js
  <p>You clicked {this.state.count} times</p>
```

在函数中，我们可以直接用 `count`:

```js
  <p>You clicked {count} times</p>
```

## 更新 State

> 在 class 中，我们需要调用 `this.setState()` 来更新 `count` state：

Reviewer 3: 这里不翻译 state 感觉有点怪，改成"更新 `count` 的值"是不是好一点。

```js{1}
  <button onClick={() => this.setState({ count: this.state.count + 1 })}>
    Click me
  </button>
```

在函数中，我们已经有了 `setCount` 和 `count` 变量，所以我们不需要 `this`:

```js{1}
  <button onClick={() => setCount(count + 1)}>
    Click me
  </button>
```

> ## 重述

Reviewer 4: 总结，回顾，复习，感觉好听一点。

> 现在让我们来**仔细回顾学到的知识**，看下我们是否真正理解了。

Reviewer 3: ```diff
- 仔细回顾学到的知识
+ 仔细回顾一下学到的知识
```

  }
```

> 在函数组件内部，我们没有 `this`，所以我们不能读或写 `this.state`。我们直接在我们组件内部调用 `useState` 钩子函数：

Reviewer 4: 同上，勾子函数

```js{4,5}
import { useState } from 'react';

function Example() {
  // 声明一个新状态变量：count
  const [count, setCount] = useState(0);
```

> ** `useState` 方法做了哪些事情 ?** 它定义一个“状态变量”。我们命名它为 `count` ， 你也可以给它取其他的名字，像 `banana` 。这是一种在函数调用过程中保存值的方式 —— `useState` 是一种使用 class 里面的 `this.state` 的新方法。一般来说，在函数退出后变量就就会"消失"，但状态变量在函数退出后仍然会被 React 保存。

Reviewer 4: ```suggestion
**useState` 方法做了哪些事情 ?** 它定义一个“状态变量”。我们命名它为 `count` ， 你也可以给它取其他的名字，例如 `banana` 。这是一种在函数调用过程中保存值的方式 —— `useState` 是一种使用 class 里面的 `this.state` 的新方法。一般来说，在函数退出后变量就就会"消失"，但状态变量在函数退出后仍然会被 React 保存。
```

Reviewer 4: 注意空格问题

**我们应该传递给 `useState` 哪些参数？** `useState()` 方法里面唯一的参数就是初始状态。与 class 不一样，状态可以不是对象。如果我们只需要数字或字符串，我们就可以只存储字符串和数字。在我们例子中，我们只要一个数字记录用户点击次数，所以我们传了 `0` 作为我们状态的初始值。（如果想要存储两个不同的状态变量，只需调用 `useState()`两次即可。）

**`useState`方法的返回值是什么？** 它返回一组值：当前状态和一个更新状态的函数。这就是我们写 `const [count, setCount] = useState()` 的原因。

> 这和 class 里面的 `this.state.count` 和 `this.setState` 类似，不过你同时返回了两个。如果你们对我们使用的语法不熟悉，你可以在[底部](/docs/hooks-state.html#tip-what-do-square-brackets-mean)找到它。

Reviewer 4: 不要另外增加多余的行。140 应该和 139 是一行

既然我们知道了 `useState` 干了什么，我们的例子应该更加容易理解了：

```js{4,5}
import { useState } from 'react';

function Example() {
  // 声明状态变量：count
  const [count, setCount] = useState(0);
```

我们声明了 `count` 变量，然后把它设为 `0`。React 会在重新渲染的时候保存它当前的值，并且把最新的值传递给我们的函数。如果我想要更新当前的 `count`,我们可以调用 `setCount`。

>注意
>
>你可能好奇：为什么叫 `useState` 而不叫 `createState`?
>

> >"Create"可能不是很准确，因为状态只在组件首次渲染的时候创建。在下一次重新渲染时，`useState` 返回给你当前的状态。否则状态就不会被"保持"了！这也是钩子函数*总是*以`use`开头的一个原因。我们也可以访问[Hooks规则](/docs/hooks-rules.html)了解这些。

Reviewer 4: 不要翻译为钩子函数，直接叫 Hooks

## 读取状态

> 当我们在 class 中显示当前的 count，我们直接访问 `this.state.count`：

Reviewer 4: ```suggestion
当需要在 class 中显示当前的 count时，可以读取 `this.state.count`：
```

```js
  <p>You clicked {this.state.count} times</p>
```

> 在函数里面，我们可以直接用 `count`:

Reviewer 4: 这里是不是多删了一行？需要校对一下

```js
  <p>You clicked {count} times</p>
```

## 更新状态

在 class 中，我们需要调用 `this.setState()` 来更新 `count` 状态：

```js{1}
  <button onClick={() => this.setState({ count: this.state.count + 1 })}>
    Click me
  </button>
```

在函数中，我们已经有了 `setCount` 和 `count` 变量，所以我们不需要 `this`:

```js{1}
  <button onClick={() => setCount(count + 1)}>
    Click me
  </button>
```

## 总结

现在我们可以**一行行过一下**代码，并且看下我们是否真正理解了。

<!--
  I'm not proud of this line markup. Please somebody fix this.
  But if GitHub got away with it for years we can cheat.
-->
```js{1,4,9}
 1:  import { useState } from 'react';

>  2: 

Reviewer 4: 多一个空格

14:  }
```

* **第一行:** 引入 React 中的 `useState` Hook。它让我们在函数组件中存储内部状态。

> * **第四行:** 在 `Example` 组件内部，我们通过调用 `useState` Hook 声明了一个新的状态变量。它返回一对值给到我们命名的变量上。我们把变量命名为 `count`,因为它存储的是点击次数。我们通过传 `0` 作为 `useState` 唯一参数来将其初始化为 `0`。第二个返回的值本身就是一个函数。它让我们可以更新 `count` 的值,所以我们叫它 `setCount`。

Reviewer 4: ```suggestion
* **第四行:** 在 `Example` 组件内部，我们通过调用 `useState` Hook 声明了一个新的状态变量。它返回一对值给到我们命名的变量上。我们把变量命名为 `count`，因为它存储的是点击次数。我们通过传 `0` 作为 `useState` 唯一参数来将其初始化为 `0`。第二个返回的值本身就是一个函数。它让我们可以更新 `count` 的值，所以我们叫它 `setCount`。
```

> * **第四行:** 在 `Example` 组件内部，我们通过调用 `useState` Hook 声明了一个新的 state 变量。它返回一对值给到我们命名的变量上。我们把变量命名为 `count`，因为它存储的是点击次数。我们通过传 `0` 作为 `useState` 唯一参数来将其初始化为 `0`。第二个返回的值本身就是一个函数。它让我们可以更新 `count` 的值，所以我们叫它 `setCount`。

Reviewer 3: ```diff
- 作为 `useState` 唯一参数
+ 作为 `useState` 唯一的参数
```

* **第九行:** 当用户点击按钮后，我们传递一个新的值给 `setCount`。React 会重新渲染 `Example` 组件，并把最新的 `count` 传给它。

乍一看这似乎有点太多了。不要急于求成！如果你有不理解的地方，请再次查看以上代码并从头到尾阅读。我们保证一旦你试着"忘记" class 里面 state 是如何工作的，并用新的眼光看这段代码，就容易理解了。 

### 提示：方括号有什么用？

你可能注意到我们用方括号定义了一个状态变量

```js
  const [count, setCount] = useState(0);
```

等号左边名字并不是 React API 的部分，你可以自己取名字：

```js
  const [fruit, setFruit] = useState('banana');
```

> 这种 Javascript 语法叫[数组解构](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Array_destructuring)。它意味着我们同时创建了 `fruit` 和 `setFruit` 两个变量，`fruit`值为 `useState` 返回的第一个值，`setFruit` 是返回的第二个值。它等价于下面的代码：

Reviewer 4: ```suggestion
这种 Javascript 语法叫[数组解构](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Array_destructuring)。它意味着我们同时创建了 `fruit` 和 `setFruit` 两个变量，`fruit` 值为 `useState` 返回的第一个值，`setFruit` 是返回的第二个值。它等价于下面的代码：
```

> 这种 Javascript 语法叫[数组解构](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Array_destructuring)。它意味着我们同时创建了 `fruit` 和 `setFruit` 两个变量，`fruit` 值为 `useState` 返回的第一个值，`setFruit` 是返回的第二个值。它等价于下面的代码：

Reviewer 3: `JavaScript`

Reviewer 3: ```diff
- `fruit` 值
+ `fruit` 的值
```

```js

>   var fruitStateVariable = useState('banana'); // 返回一个数组

Reviewer 4: ```suggestion
  var fruitStateVariable = useState('banana'); // 返回一个有两个元素的数组
```

Reviewer 4: 感觉 pair 的意思需要表达出来

  var fruit = fruitStateVariable[0]; // 数组里的第一个值
  var setFruit = fruitStateVariable[1]; // 数组里的第二个值
```

> 当我们使用 `useSatate` 定义 一个 state 变量时候，它返回一个有两个值的数组。第一个值是当前的 state，第二值是更新 state 的函数。使用 `[0]` 和 `[1]` 来访问有点令人困惑，因为它们有特定的含义。这就是我们使用数组解构的原因。

Reviewer: ```suggestion
当我们使用 `useSatate` 定义 一个 state 变量时候，它返回一个有两个值的数组。第一个值是当前的 state，第二个值是更新 state 的函数。使用 `[0]` 和 `[1]` 来访问有点令人困惑，因为它们有特定的含义。这就是我们使用数组解构的原因。
```

> 当我们使用 `useSatate` 定义 一个 state 变量时候，它返回一对值——有两个值的数组。第一个值是当前的 state，第二值是更新 state 的函数。使用 `[0]` 和 `[1]` 来访问有点令人困惑，因为它们有特定的含义。这就是我们使用数组析构的原因。

Reviewer 3: ```diff
- 它返回一对值——有两个值的数组。
+ 它返回一个有两个值的数组。
```

Reviewer 3: ```diff
- 数组析构
+ 数组解构
```

>注意
>

> >你可能会好奇React怎么知道 `useState` 对应的是哪个组件，因为我们并没有传递 `this` 给 React。我们将在 FAQ 部分回答[这个问题](/docs/hooks-faq.html#how-does-react-associate-hook-calls-with-components)以及许多其他问题。

Reviewer 4: ```suggestion
>你可能会好奇 React 怎么知道 `useState` 对应的是哪个组件，因为我们并没有传递 `this` 给 React。我们将在 FAQ 部分回答[这个问题](/docs/hooks-faq.html#how-does-react-associate-hook-calls-with-components)以及许多其他问题。
```

> ### 提示：使用多个状态变量

Reviewer 4: ```suggestion
### 提示：使用多个 state 变量
```

将 state 变量声明为一对 `[something, setSomething]` 也很方便，因为如果我们想使用多个 state 变量，它允许我们给不同的 state 变量取不同的名称：

```js
function ExampleWithManyStates() {
  // 声明多个 state 变量
  const [age, setAge] = useState(42);
  const [fruit, setFruit] = useState('banana');
  const [todos, setTodos] = useState([{ text: 'Learn Hooks' }]);
```

在以上组件中，我们有局部变量 `age`,`fruit` 和 `todos`，并且我们可以单独更新它们：

```js
  function handleOrangeClick() {
    // 和 this.setState({ fruit: 'orange' }) 类似
    setFruit('orange');
  }
```

> 你**不必**使用多个状态变量。状态变量可以很好的存储对象和数组，因此，你仍然可以将相关数据分为一组。然而，不像 class 中的 `this.setState`，更新 state 变量总是`替换`它而不是合并它。

Reviewer 4: ```suggestion
你**不必**使用多个 state 变量。State 变量可以很好的存储对象和数组，因此，你仍然可以将相关数据分为一组。然而，不像 class 中的 `this.setState`，更新 state 变量总是*替换*它而不是合并它。
```

Reviewer 4: 替换两边是 asterisk

> 我们提供了更多关于分离独立状态变量的建议 [在 FAQ 中](/docs/hooks-faq.html#should-i-use-one-or-many-state-variables)。

Reviewer 4: ```suggestion
我们[在 FAQ 中](/docs/hooks-faq.html#should-i-use-one-or-many-state-variables)提供了更多关于分离独立状态变量的建议 。
```

> 我们[在 FAQ 中](/docs/hooks-faq.html#should-i-use-one-or-many-state-variables)提供了更多关于分离独立 state 变量的建议 。

Reviewer 2: ```suggestion
我们[在 FAQ 中](/docs/hooks-faq.html#should-i-use-one-or-many-state-variables)提供了更多关于分离独立 state 变量的建议。
```

## 下一步

> 上述页面中，我们了解了 React 提供的一个叫 `useState` Hook，有时候我们也叫它 “State Hook”。它让我们在 React 函数定义组件上添加内部 state——这是我们第一次这么做。

Reviewer: "did for the first time ever" 有“破天荒头一回”的意思，感情强烈最好体现一下。抛砖引玉：这是我们头一次能这么干！

> 我们也学到了一点点关于 Hook 是什么的知识。Hook 是能让你在函数组件中“钩住”React特性的函数。它们名字通常都以 `use` 开始，还有更多 Hook 等着我们去探索。

Reviewer 5: 1. 我们还学到了一些知识比如什么是 Hook。
2. 中英文空格问题。

> 我们还学到了一些知识比如什么是 Hook。Hook 是能让你在函数定义组件中“钩住” React 特性的函数。它们名字通常都以 `use` 开始，还有更多 Hook 等着我们去探索。

Reviewer: 全文保持一致

```suggestion
我们还学到了一些知识比如什么是 Hook。Hook 是能让你在函数定义组件中“钩入” React 特性的函数。它们名字通常都以 `use` 开始，还有更多 Hook 等着我们去探索。
```

> **现在我们继续下一章[学习下一个 Hook: `useEffect` .](/docs/hooks-effect.html)** 它让你产生组件中的副作用，并且它跟 class 里面的生命周期函数很类似。

Reviewer 4: ```suggestion
**现在我们继续下一章[学习下一个 Hook: `useEffect` 。](/docs/hooks-effect.html)** 它让你实现组件中的副作用，并且它跟 class 里面的生命周期函数很类似。
```

> **现在我们继续下一章[学习下一个 Hook: `useEffect` ](/docs/hooks-effect.html)。** 它让你产生组件中的副作用，并且它跟 class 里面的生命周期函数很类似。

Reviewer: ```suggestion
**现在我们继续下一章[学习下一个 Hook: `useEffect` ](/docs/hooks-effect.html)。** 它让你能在组件中产生副作用，并且它跟 class 里面的生命周期函数很类似。
```

> **现在我们继续下一章[学习下一个 Hook: `useEffect` 。](/docs/hooks-effect.html)** 它让你产生组件中的副作用，并且它跟 class 里面的生命周期函数很类似。

Reviewer 3: 句号应该在中括号外面， 原文也有误。

```diff
- 它让你产生组件中的副作用
+ 它让你在组件处理副作用
```

14:  }
```

> * **第一行:** 引入 React 中的`useState` 钩子函数。它让我们在函数组件中存储内部状态。

Reviewer 4: ditto，~勾子函数~ -> Hook

> * **第四行:** 在 `Example` 组件内部，我们调用 `useState` 钩子函数定义了一个新的状态变量。它返回一个数组给我们的变量。我们把变量命名为 `count`,因为它存储的是点击次数。我们通过传参数`0`给 `useState` 方法来把变量初始化为 `0`。第二个返回的值是一个函数。它让我们可以更新 `count` 的值,所以我们叫它 `setCount`。

Reviewer 4: ditto

你可能感觉要理解的东西太多了。不要急于求成！如果你有不理解的地方，再看一遍上面的代码并且从头读到尾。我们保证只要你试着"忘记" class 里面 state 的工作原理，并且擦亮眼睛看下这段代码，就会有种豁然开朗的感觉。 

### 提示：方括号有什么用？

你可能注意到我们用方括号定义了一个状态变量

```js
  const [count, setCount] = useState(0);
```

等号左边名字并不是 React API 的部分，你可以自己取名字：

```js
  const [fruit, setFruit] = useState('banana');
```

这种 Javascript 语法叫[数组解构](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Array_destructuring)。它意味着我们同时创建了 `fruit` 和 `setFruit` 两个变量，
`fruit`值为 `useState` 返回的第一个值，`setFruit` 是返回的第二个。它等价于下面的代码：

```js
  var fruitStateVariable = useState('banana'); // 返回一个数组
  var fruit = fruitStateVariable[0]; // 数组里的第一个值
  var setFruit = fruitStateVariable[1]; // 数组里的第二个值
```

当我们使用 `useSatate` 定义变量时候，它返回有两个值的数组。第一个是当前的状态，第二个是更新状态的函数。使用 `[0]` 和 `[1]` 来访问它们会让人困惑，因为他们的含义不一样。这就是为什么我们使用数组析构的方式的原因。

>注意
>
>你可能会好奇React怎么知道是哪个组件的 `useState` ，因为我们并没有传递 `this` 给 React 。我们将在 FAQ 里面对[这些问题](/docs/hooks-faq.html#how-does-react-associate-hook-calls-with-components)及其他问题进行解答。

### 提示：使用多个状态变量

使用 `[something, setSomething]` 方式定义状态变量会很便捷，因为如果我们需要使用不止一个状态变量，我们可以使用不同的名字：

```js
function ExampleWithManyStates() {

>   // 定义多个状态变量

Reviewer 4: ```suggestion
  // 声明多个状态变量
```
FYI，定义是 define

  const [age, setAge] = useState(42);
  const [fruit, setFruit] = useState('banana');
  const [todos, setTodos] = useState([{ text: 'Learn Hooks' }]);
```
上面组件中，我们有 `age`,`fruit` 和 `todos` 局部变量，并且我们可以独立地更新它们：

```js
  function handleOrangeClick() {
    // 和 this.setState({ fruit: 'orange' }) 类似
    setFruit('orange');
  }
```

> 你 **不一定** 要使用多个状态变量。状态变量也可以存储对象和数组，所以你可以把有关联的数据组合起来。然而，不像在class内部`this.setState`，更新一个状态变量往往`替换`它而不是合并它。

Reviewer 4: class 两边空格
然后我觉得这一句话读起来不太通顺，可以整理一下

关于分割独立的状态变量，我们在[FAQ](/docs/hooks-faq.html#should-i-use-one-or-many-state-variables)中提供了很多建议。

## 下一步

> 上述页面中，我们已经学到了React提供的一个叫 `useState` 钩子函数，有时候我们也叫它 `State Hook`。它让我们在React函数组件上添加局部状态，局部状态只需初始化一次就可一直使用。

Reviewer 4: ditto，钩子

Reviewer 4: React 两边空格

Reviewer 4: 规范里面叫 “函数定义组件”

> 我们也学到了一点点关于 Hooks 是什么的知识。Hooks 是能让你在函数组件中接入React特性的函数。它们名字都以 `use` 开始，还有很多 Hooks 等着我们去探索。

Reviewer 4: React 两边空格

> **接下来我们继续下一章[学习下一个Hook: `useEffect` .](/docs/hooks-effect.html)** 它让你了解到组件的副作用，并且它跟class里面的生命周期函数很类似。

Reviewer 4: class 两边空格
